File: programmers/level 2/숫자_카드_나누기.py
# 240618
def find_gcd(num1, num2):
    while num1:
        num1, num2 = num2 % num1, num1
    return num2


def solution(arrayA, arrayB):
    length = len(arrayA)
    if set(arrayA) & set(arrayB):
        return 0
    if length == 1:
        return max(*arrayA, *arrayB)

    arrayA.sort()
    arrayB.sort()
    gcd1 = find_gcd(*arrayA[:2])

    for i in range(2, length):
        gcd1 = find_gcd(gcd1, arrayA[i])

    gcd2 = find_gcd(*arrayB[:2])

    for i in range(2, length):
        gcd2 = find_gcd(gcd2, arrayB[i])

    if gcd1 == gcd2:
        return 0

    gcd = 0
    for i in range(length):
        if arrayB[i] % gcd1 == 0:
            break
    else:
        gcd = gcd1

    for i in range(length):
        if arrayA[i] % gcd2 == 0:
            break
    else:
        gcd = max(gcd, gcd2)

    return gcd

File: programmers/level 2/숫자_카드_나누기.js
// 250717
function solution(...arr) {
    const n = arr[0].length
    let answer = 0

    arr.forEach((element, idx) => {
        element.sort((a, b) => b-a)
        let gcd = element[0]
        for (let i = 1; i < n; i += 1) {
            let num = element[i]
            while (num) {
                [gcd, num] = [num, gcd % num]
            }

        }

        let possible = true
        for (let j = 0; j < n; j += 1) {
            if (arr[1-idx][j] % gcd === 0) {
                possible = false
                break
            }
        }
        if (possible) {
            answer = Math.max(answer, gcd)
        }

    })


    return answer
}